File: packages/dart/lib/src/merge.dart
import 'aad.dart';
import 'errors.dart';
import 'registries.dart';
import 'types.dart';

MergeResult mergePayloads(
  VaultPayload a,
  VaultPayload b,
  List<UnlockSlot> slotsA,
  List<UnlockSlot> slotsB,
) {
  if (a.identity.identityId != b.identity.identityId) {
    fail('ERR_IDENTITY_MISMATCH');
  }
  if (a.identity.type != b.identity.type || a.identity.value != b.identity.value) {
    fail('ERR_IDENTITY_MISMATCH');
  }

  final conflicts = <MergeConflict>[];
  if (a.msk.current.mskId != b.msk.current.mskId) {
    fail('ERR_MERGE_MSK', 'current MSK mismatch is a hard conflict');
  }

  final history = _unionHistory(a.msk.history, b.msk.history);
  final keys = _mergeKeys(a.keys, b.keys, a.tombstones, b.tombstones, conflicts);
  final preferred = _mergePreferred(a.preferredKeys, b.preferredKeys, keys, conflicts);
  final tombstones = _unionTombstones(a.tombstones, b.tombstones);
  final slots = _mergeSlots(slotsA, slotsB, conflicts);
  final created = a.metadata.createdAt.compareTo(b.metadata.createdAt) <= 0
      ? a.metadata.createdAt
      : b.metadata.createdAt;
  final updated = a.metadata.updatedAt.compareTo(b.metadata.updatedAt) >= 0
      ? a.metadata.updatedAt
      : b.metadata.updatedAt;

  final payload = VaultPayload(
    identity: a.identity,
    msk: MskState(current: a.msk.current, history: history),
    keys: keys,
    preferredKeys: preferred,
    metadata: VaultMetadata(createdAt: created, updatedAt: updated),
    tombstones: tombstones,
    extensions: _unionExtensions(a.extensions, b.extensions, conflicts),
    criticalExtensions:
        _unionExtensions(a.criticalExtensions, b.criticalExtensions, conflicts),
  );

  if (payload.criticalExtensions.isNotEmpty) {
    fail('ERR_CRITICAL_EXTENSION', 'unknown critical extension during merge');
  }

  return MergeResult(payload: payload, slots: slots, conflicts: conflicts);
}

List<KeyRecord> _mergeKeys(
  List<KeyRecord> a,
  List<KeyRecord> b,
  List<Tombstone> tombsA,
  List<Tombstone> tombsB,
  List<MergeConflict> conflicts,
) {
  final map = <String, KeyRecord>{};
  for (final k in a) {
    map[k.absoluteKeyId] = k.copyWith();
  }
  for (final k in b) {
    final existing = map[k.absoluteKeyId];
    if (existing == null) {
      map[k.absoluteKeyId] = k.copyWith();
      continue;
    }
    final sevA = statusSeverity[existing.status] ?? 0;
    final sevB = statusSeverity[k.status] ?? 0;
    final status = sevA >= sevB ? existing.status : k.status;
    String? privateKey = existing.privateKey ?? k.privateKey;
    final tombA = _tombstoneFor(tombsA, k.absoluteKeyId);
    final tombB = _tombstoneFor(tombsB, k.absoluteKeyId);
    var clear = false;
    if (tombA != null && tombB != null && tombA.nonce == tombB.nonce) {
      clear = true;
      privateKey = null;
    } else if ((tombA != null && tombB == null && k.privateKey != null) ||
        (tombB != null && tombA == null && existing.privateKey != null)) {
      conflicts.add(
        MergeConflict(
          code: 'ERR_MERGE_PRIVATE_KEY',
          message: 'destructive delete conflict for ${k.absoluteKeyId}',
        ),
      );
      privateKey = existing.privateKey ?? k.privateKey;
    } else {
      privateKey = existing.privateKey ?? k.privateKey;
    }
    final created = existing.createdAt.compareTo(k.createdAt) <= 0
        ? existing.createdAt
        : k.createdAt;
    map[k.absoluteKeyId] = existing.copyWith(
      status: status,
      privateKey: privateKey,
      createdAt: created,
      clearPrivateKey: clear,
    );
  }
  final keys = map.values.toList()
    ..sort((x, y) => x.absoluteKeyId.compareTo(y.absoluteKeyId));
  return keys;
}

Map<String, Map<String, String>> _mergePreferred(
  Map<String, Map<String, String>> a,
  Map<String, Map<String, String>> b,
  List<KeyRecord> keys,
  List<MergeConflict> conflicts,
) {
  final out = <String, Map<String, String>>{};
  final families = {...a.keys, ...b.keys};
  for (final family in families) {
    final pa = a[family] ?? const <String, String>{};
    final pb = b[family] ?? const <String, String>{};
    final purposes = {...pa.keys, ...pb.keys};
    for (final purpose in purposes) {
      final chosen = _pickPreferred(
        pa[purpose],
        pb[purpose],
        keys,
        conflicts,
        family,
        purpose,
      );
      if (chosen != null) {
        out.putIfAbsent(family, () => <String, String>{})[purpose] = chosen;
      }
    }
  }
  return out;
}

String? _pickPreferred(
  String? va,
  String? vb,
  List<KeyRecord> keys,
  List<MergeConflict> conflicts,
  String family,
  String purpose,
) {
  bool valid(String? id) =>
      id != null && keys.any((k) => k.absoluteKeyId == id);
  if (va != null && vb != null && va != vb) {
    conflicts.add(
      MergeConflict(code: 'ERR_MERGE_PREFERRED_KEY', message: '$family/$purpose'),
    );
    return null;
  }
  final id = va ?? vb;
  return valid(id) ? id : null;
}

List<UnlockSlot> _mergeSlots(
  List<UnlockSlot> a,
  List<UnlockSlot> b,
  List<MergeConflict> conflicts,
) {
  final map = <String, UnlockSlot>{};
  for (final s in a) {
    map[s.slotId] = s;
  }
  for (final s in b) {
    final existing = map[s.slotId];
    if (existing == null) {
      map[s.slotId] = s;
      continue;
    }
    if (slotFingerprint(existing) != slotFingerprint(s)) {
      conflicts.add(MergeConflict(code: 'ERR_MERGE_SLOT', message: s.slotId));
      fail('ERR_MERGE_SLOT', 'slot ${s.slotId} differs');
    }
  }
  return map.values.toList();
}

Tombstone? _tombstoneFor(List<Tombstone> tombs, String id) {
  for (final t in tombs) {
    if (t.absoluteKeyId == id) return t;
  }
  return null;
}

List<MskHistoryEntry> _unionHistory(
  List<MskHistoryEntry> a,
  List<MskHistoryEntry> b,
) {
  final map = <String, MskHistoryEntry>{};
  for (final h in [...a, ...b]) {
    map[h.mskId] = h;
  }
  return map.values.toList();
}

List<Tombstone> _unionTombstones(List<Tombstone> a, List<Tombstone> b) {
  final map = <String, Tombstone>{};
  for (final t in [...a, ...b]) {
    map['${t.absoluteKeyId}:${t.nonce}'] = t;
  }
  return map.values.toList();
}

List<Extension> _unionExtensions(
  List<Extension> a,
  List<Extension> b,
  List<MergeConflict> conflicts,
) {
  final map = <String, Extension>{};
  for (final e in [...a, ...b]) {
    final existing = map[e.id];
    if (existing != null &&
        existing.data.toString() != e.data.toString()) {
      conflicts.add(
        MergeConflict(
          code: 'ERR_EXTENSION',
          message: 'extension data conflict ${e.id}',
        ),
      );
      continue;
    }
    map[e.id] = e;
  }
  return map.values.toList();
}

File: packages/js/packages/core/src/merge.ts
import { fail } from "./errors.js";
import { STATUS_SEVERITY } from "./registries.js";
import type {
  KeyRecord,
  MergeConflict,
  MergeResult,
  MskHistoryEntry,
  PreferredKeys,
  Tombstone,
  UnlockSlot,
  VaultPayload,
} from "./types.js";
import { slotFingerprint } from "./aad.js";

export function mergePayloads(a: VaultPayload, b: VaultPayload, slotsA: UnlockSlot[], slotsB: UnlockSlot[]): MergeResult {
  if (a.identity.identity_id !== b.identity.identity_id) fail("ERR_IDENTITY_MISMATCH");
  if (a.identity.type !== b.identity.type || a.identity.value !== b.identity.value) fail("ERR_IDENTITY_MISMATCH");

  const conflicts: MergeConflict[] = [];

  if (a.msk.current.msk_id !== b.msk.current.msk_id) {
    fail("ERR_MERGE_MSK", "current MSK mismatch is a hard conflict");
  }

  const history = unionHistory(a.msk.history, b.msk.history);
  const keys = mergeKeys(a.keys, b.keys, a.tombstones, b.tombstones, conflicts);
  const preferred = mergePreferred(a.preferred_keys, b.preferred_keys, keys, conflicts);
  const tombstones = unionTombstones(a.tombstones, b.tombstones);
  const slots = mergeSlots(slotsA, slotsB, conflicts);
  const created = a.metadata.created_at <= b.metadata.created_at ? a.metadata.created_at : b.metadata.created_at;

  const payload: VaultPayload = {
    identity: a.identity,
    msk: { current: a.msk.current, history },
    keys,
    preferred_keys: preferred,
    metadata: {
      created_at: created,
      updated_at: a.metadata.updated_at >= b.metadata.updated_at ? a.metadata.updated_at : b.metadata.updated_at,
    },
    tombstones,
    extensions: unionExtensions(a.extensions, b.extensions, conflicts),
    critical_extensions: unionExtensions(a.critical_extensions, b.critical_extensions, conflicts),
  };

  if (payload.critical_extensions.length) {
    fail("ERR_CRITICAL_EXTENSION", "unknown critical extension during merge");
  }

  return { payload, slots, conflicts };
}

function mergeKeys(
  a: KeyRecord[],
  b: KeyRecord[],
  tombsA: Tombstone[],
  tombsB: Tombstone[],
  conflicts: MergeConflict[],
): KeyRecord[] {
  const map = new Map<string, KeyRecord>();
  for (const k of a) map.set(k.absolute_key_id, { ...k });
  for (const k of b) {
    const existing = map.get(k.absolute_key_id);
    if (!existing) {
      map.set(k.absolute_key_id, { ...k });
      continue;
    }
    const sevA = STATUS_SEVERITY[existing.status] ?? 0;
    const sevB = STATUS_SEVERITY[k.status] ?? 0;
    const status = sevA >= sevB ? existing.status : k.status;
    let private_key = existing.private_key ?? k.private_key;
    const tombA = tombsA.find((t) => t.absolute_key_id === k.absolute_key_id);
    const tombB = tombsB.find((t) => t.absolute_key_id === k.absolute_key_id);
    if (tombA && tombB && tombA.nonce === tombB.nonce) {
      private_key = null;
    } else if ((tombA && !tombB && k.private_key) || (tombB && !tombA && existing.private_key)) {
      conflicts.push({
        code: "ERR_MERGE_PRIVATE_KEY",
        message: `destructive delete conflict for ${k.absolute_key_id}`,
      });
      private_key = existing.private_key ?? k.private_key;
    } else {
      private_key = existing.private_key ?? k.private_key;
    }
    map.set(k.absolute_key_id, {
      ...existing,
      status,
      private_key,
      created_at: existing.created_at <= k.created_at ? existing.created_at : k.created_at,
    });
  }
  return [...map.values()].sort((x, y) => x.absolute_key_id.localeCompare(y.absolute_key_id));
}

function mergePreferred(
  a: PreferredKeys,
  b: PreferredKeys,
  keys: KeyRecord[],
  conflicts: MergeConflict[],
): PreferredKeys {
  const out: PreferredKeys = {};
  const families = new Set([...Object.keys(a), ...Object.keys(b)]) as Set<keyof PreferredKeys>;
  for (const family of families) {
    const pa = a[family] ?? {};
    const pb = b[family] ?? {};
    const purposes = new Set([...Object.keys(pa), ...Object.keys(pb)]);
    for (const purpose of purposes) {
      const va = pa[purpose as keyof typeof pa];
      const vb = pb[purpose as keyof typeof pb];
      const chosen = pickPreferred(va, vb, keys, conflicts, String(family), String(purpose));
      if (chosen) {
        out[family] ??= {};
        (out[family] as Record<string, string>)[purpose] = chosen;
      }
    }
  }
  return out;
}

function pickPreferred(
  va: string | undefined,
  vb: string | undefined,
  keys: KeyRecord[],
  conflicts: MergeConflict[],
  family: string,
  purpose: string,
): string | undefined {
  const valid = (id?: string) => Boolean(id && keys.some((k) => k.absolute_key_id === id && k.private_key !== undefined));
  if (va && vb && va !== vb) {
    conflicts.push({ code: "ERR_MERGE_PREFERRED_KEY", message: `${family}/${purpose}` });
    return undefined;
  }
  const id = va ?? vb;
  return valid(id) ? id : undefined;
}

function mergeSlots(a: UnlockSlot[], b: UnlockSlot[], conflicts: MergeConflict[]): UnlockSlot[] {
  const map = new Map<string, UnlockSlot>();
  for (const s of a) map.set(s.slot_id, s);
  for (const s of b) {
    const existing = map.get(s.slot_id);
    if (!existing) {
      map.set(s.slot_id, s);
      continue;
    }
    if (slotFingerprint(existing) !== slotFingerprint(s)) {
      conflicts.push({ code: "ERR_MERGE_SLOT", message: s.slot_id });
      fail("ERR_MERGE_SLOT", `slot ${s.slot_id} differs`);
    }
  }
  return [...map.values()];
}

function unionHistory(a: MskHistoryEntry[], b: MskHistoryEntry[]): MskHistoryEntry[] {
  const map = new Map<string, MskHistoryEntry>();
  for (const h of [...a, ...b]) map.set(h.msk_id, h);
  return [...map.values()];
}

function unionTombstones(a: Tombstone[], b: Tombstone[]): Tombstone[] {
  const map = new Map<string, Tombstone>();
  for (const t of [...a, ...b]) map.set(`${t.absolute_key_id}:${t.nonce}`, t);
  return [...map.values()];
}

function unionExtensions(
  a: VaultPayload["extensions"],
  b: VaultPayload["extensions"],
  conflicts: MergeConflict[],
): VaultPayload["extensions"] {
  const map = new Map<string, VaultPayload["extensions"][number]>();
  for (const e of [...a, ...b]) {
    const existing = map.get(e.id);
    if (existing && JSON.stringify(existing.data) !== JSON.stringify(e.data)) {
      conflicts.push({ code: "ERR_EXTENSION", message: `extension data conflict ${e.id}` });
      continue;
    }
    map.set(e.id, e);
  }
  return [...map.values()];
}
